test/correctness.cpp

```diff
@@ -0,0 +1,422 @@
+#include <gtest/gtest.h>
+#include <vector>
+#include <format>
+
+#include <piwcsprwmodel.h>
+
+#include "../src/nodetypeinfo.h"
+#include "../src/correctness.h"
+
+using namespace piwcs::prw;
+
+using SType = Section::AllowedTravel;
+
+struct SectionTestInfo{
+    Section::AllowedTravel type;
+    bool inward = false; // Section is directed into node if `true`
+};
+
+template<std::size_t N>
+bool localCorrectnessMinimal(
+        NodeType type,
+        const SectionTestInfo (& sections)[N]
+        ){
+
+    assert(
+            N>=(type->slotCount) &&
+            "Invalid section array provided, slot counts don't match!"
+            );
+
+    Model model;
+    model.newNode(type, "n0");
+
+    for(SlotId slot=0; slot<(type->slotCount); slot++){
+        Identifier nodeId = std::format("n{}", slot+1);
+        Identifier sectionId = std::format("s{}", slot);
+        model.newNode(type, nodeId);
+        model.newSection(sectionId, sections[slot].type, nullptr);
+
+        if(sections[slot].inward){
+            model.link(sectionId, nodeId, 0, "n0", slot);
+        }
+        else{
+            model.link(sectionId, "n0", slot, nodeId, 0);
+        }
+    }
+
+    return isLocallyCorrect(model, "n0");
+}
+
+#define UNWRAP(...) __VA_ARGS__
+#define INCORRECT_TEST(testName, nodeType, sectionInfo)             \
+    TEST(LocalCorrectness, nodeType ##_## testName ## _IsIncorrect){\
+        bool res = localCorrectnessMinimal(                         \
+                nodeType,                                           \
+                UNWRAP sectionInfo                                  \
+            );                                                      \
+   ASSERT_EQ(res, false);                                           \
+    }                                                               \
+
+#define CORRECT_TEST(testName, nodeType, sectionInfo)               \
+    TEST(LocalCorrectness, nodeType ##_## testName ## _IsCorrect){  \
+        bool res = localCorrectnessMinimal(                         \
+                nodeType,                                           \
+                UNWRAP sectionInfo                                  \
+            );                                                      \
+        ASSERT_EQ(res, true);                                       \
+    }                                                               \
+
+// << "Expected correct node of type `" << #nodeType << "`";
+
+std::vector<NodeType> types = {
+        END,
+        THRU,
+        CROSSING,
+        MOTORIZED,
+        FIXED,
+        PASSIVE,
+        MANUAL
+};
+
+TEST(LocalCorrectness, AnyCompletelyForbidden_IsCorrect){
+
+    SectionTestInfo sections[Node::MAX_SLOTS];
+
+    for (SlotId slot=0; slot<Node::MAX_SLOTS; ++slot){
+        sections[slot] = {SType::NONE};
+    }
+
+    for(NodeType type : types){
```

Comment:
Inline `types` here

Reply:
My reasoning behind `types` being a variable was that we could need some sort of list of every possible `NodeType`. It turns out such list is already inlined in [`io_read.cpp:48`](https://github.com/OLEGSHA/PiwcsPrwModel/blob/ab09de4b3232a70d08bee4840935827ab4a69a50/src/io_read.cpp#L48C29-L56C11). What are your thoughts on extracting it to a `const` variable in `nodes.h` where `NodeType` is declared?

Reply:
I inlined it (b379baf875cf048a901616f6d818882849a23027) for the time being.